pkg/net/libp2p/libp2p_transport.go

```diff
@@ -0,0 +1,500 @@
+		return err
+	}
+	if errors.As(err, &streamErr) {
+		t.connWg.Add(1)
+		go t.connectToNode(ctx, dest, t.connWg)
+		return err
+	}
+
+	outBytes, err := proto.Marshal(payload)
+	if err != nil {
+		return fmt.Errorf("failed to marshal message: %w", err)
+	}
+
+	msg := TransportMessage{t.ownID.Pb(), outBytes}
+	buf := new(bytes.Buffer)
+	err = msg.MarshalCBOR(buf)
+	if err != nil {
+		return fmt.Errorf("failed to CBOR marshal message: %w", err)
+	}
+
+	if ctx.Err() != nil {
+		return err
+	}
+
+	w := bufio.NewWriter(s)
+	_, err = w.Write(buf.Bytes())
+	if err == nil {
+		err = w.Flush()
+	}
+	// If we cannot write to the stream, or we don't have a connection to that peer then try to
+	// connect to the peer.
+	if err != nil || len(t.host.Network().ConnsToPeer(s.Conn().RemotePeer())) == 0 {
+		t.connWg.Add(1)
+		go t.connectToNode(ctx, dest, t.connWg)
+		return fmt.Errorf("%s failed to write data to stream for %s: reopening", t.ownID, dest)
+	}
+
+	return nil
+}
+
+func (t *Transport) mirHandler(s network.Stream) {
+	t.logger.Log(logging.LevelDebug, "mir handler started", "src", t.ownID, "dst", s.ID())
+	defer t.logger.Log(logging.LevelDebug, "mir handler stopped", "src", t.ownID, "dst", s.ID())
+
+	defer func() {
+		t.logger.Log(logging.LevelDebug, "mir handler is closing stream", "src", t.ownID, "dst", s.ID())
+		err := s.Close()
+		if err != nil {
+			t.logger.Log(logging.LevelError, "closing stream", "src", t.ownID, "dst", s.ID(), "err", err)
+		}
+	}() // nolint
+
+	for {
+		var msg TransportMessage
+		err := msg.UnmarshalCBOR(s)
+		if err != nil {
+			t.logger.Log(logging.LevelError, "failed to read mir request", "src", t.ownID, "dst", s.ID(), "err", err)
+			return
+		}
+
+		var payload messagepb.Message
+
+		if err := proto.Unmarshal(msg.Payload, &payload); err != nil {
+			t.logger.Log(logging.LevelError, "failed to unmarshall mir request", "src", t.ownID, "dst", s.ID(), "err", err)
+			return
+		}
+
+		t.incomingMessages <- events.ListOf(
+			events.MessageReceived(types.ModuleID(payload.DestModule), types.NodeID(msg.Sender), &payload),
+		)
+	}
+}
+
+func (t *Transport) addOutboundStream(nodeID types.NodeID, s network.Stream) {
+	t.nodesLock.Lock()
+	defer t.nodesLock.Unlock()
+
+	node, found := t.nodes[nodeID]
+	if !found {
+		t.logger.Log(logging.LevelWarn, "addOutboundStream: failed to find the node", "src", t.ownID, "dst", nodeID)
+		return
+	}
+	node.Stream = s
+}
+
+func (t *Transport) streamExists(nodeID types.NodeID) bool {
+	t.nodesLock.RLock()
+	defer t.nodesLock.RUnlock()
+
+	v, found := t.nodes[nodeID]
+	return found && v.Stream != nil
+}
+
+func (t *Transport) getNodeStream(nodeID types.NodeID) (network.Stream, error) {
+	t.nodesLock.RLock()
+	defer t.nodesLock.RUnlock()
+
+	_, found := t.nodes[nodeID]
+	if !found {
+		return nil, &nodeInfoError{nodeID}
+	}
+
+	node, found := t.nodes[nodeID]
+	if !found ||
+		node.Stream == nil {
+		return nil, &nilStreamError{SrcNode: t.ownID, DestNode: nodeID}
+	}
+	return node.Stream, nil
```

Comment:
As mentioned above, Streams should be locked while being written to avoid parallel accesses.

Reply:
Yes, good catch! I combined two functions and forgot to change locking